pkg/scheduler/core/division_algorithm.go

```diff
@@ -96,170 +94,69 @@ func getStaticWeightInfoList(clusters []*clusterv1alpha1.Cluster, weightList []p
-func divideRemainingReplicas(remainingReplicas int, desiredReplicaInfos map[string]int64, clusterNames []string) {
-	if remainingReplicas <= 0 {
-		return
+// dynamicDivideReplicas assigns a total number of replicas to the selected clusters by preference according to the resource.
+func dynamicDivideReplicas(state *assignState) ([]workv1alpha2.TargetCluster, error) {
+	if state.availableReplicas < state.target {
+		return nil, fmt.Errorf("clusters resources are not enough to schedule, max %d replicas are support", state.availableReplicas)
 	}
 
-	clusterSize := len(clusterNames)
-	if remainingReplicas < clusterSize {
-		for i := 0; i < remainingReplicas; i++ {
-			desiredReplicaInfos[clusterNames[i]]++
-		}
-	} else {
-		avg, residue := remainingReplicas/clusterSize, remainingReplicas%clusterSize
-		for i := 0; i < clusterSize; i++ {
-			if i < residue {
-				desiredReplicaInfos[clusterNames[i]] += int64(avg) + 1
-			} else {
-				desiredReplicaInfos[clusterNames[i]] += int64(avg)
+	switch state.strategyType {
+	case AggregatedStrategy:
+		state.availableClusters = state.resortClusterList()
+		var sum int32
+		for i := range state.availableClusters {
+			if sum += state.availableClusters[i].Replicas; sum >= state.target {
+				state.availableClusters = state.availableClusters[:i+1]
+				break
 			}
 		}
+		fallthrough
+	case DynamicWeightStrategy:
+		// Set the availableClusters as the weight, scheduledClusters as init result, target as the dispenser object.
+		// After dispensing, the target cluster will be the combination of init result and weighted result for target replicas.
+		weightList := getStaticWeightInfoListByTargetClusters(state.availableClusters)
+		disp := newDispenser(state.target, state.scheduledClusters)
+		disp.takeByWeight(weightList)
+		return disp.result, nil
+	default:
+		// should never happen
+		return nil, fmt.Errorf("undefined strategy type: %s", state.strategyType)
 	}
 }
 
-func scaleDownScheduleByReplicaDivisionPreference(
-	spec *workv1alpha2.ResourceBindingSpec,
-	preference policyv1alpha1.ReplicaDivisionPreference,
-) ([]workv1alpha2.TargetCluster, error) {
+func dynamicScaleDown(state *assignState) ([]workv1alpha2.TargetCluster, error) {
 	// The previous scheduling result will be the weight reference of scaling down.
 	// In other words, we scale down the replicas proportionally by their scheduled replicas.
-	return divideReplicasByPreference(spec.Clusters, spec.Replicas, preference, sets.NewString())
+	// Now:
+	// 1. target is set to desired replicas.
+	// 2. availableClusters is set to the former schedule result.
+	// 3. scheduledClusters and assignedReplicas are not set, which implicates we consider this action as a first schedule.
+	state.target = state.object.Replicas
+	state.scheduledClusters = nil
+	state.buildAvailableClusters(func(_ []*clusterv1alpha1.Cluster, spec *workv1alpha2.ResourceBindingSpec) []workv1alpha2.TargetCluster {
```

Comment:
I think we probably need to sort the clusters, like: [1 replicas, 2 replicas, 3 replicas], to ensure scaling down the cluster with fewer replicas first.
The default `sort.Sort(TargetClustersList(clusterAvailableReplicas))` will sort the clusters like [3 replicas, 2 replicas, 1 replicas].
What do you think?

Reply:
Actually, we scale the clusters by weight of scheduled result. I think sort does not have any impact on scaling down according to the code in `dynamicDivideReplicas`.

Reply:
Sorry, I forgot to say that using `AggregatedStrategy`, it probably has effects.

Reply:
I guess you mean the order of available clusters has the impact on `resortClusterList`. But we already set `state.scheduledClusters = nil` so that resort does not take effect. 

Scale down is sort of re-allocation for replicas. We set available clusters as `spec.clusters` and remove scheduled results, so we could scale down these clusters by weight of them. This is sort of trick, I think.

Reply:
For example, the previous schedule result is {cluster-1: 4, cluster-2: 8}, available replicas is {cluster-1: 100, cluster-2: 100}, now we change the replicas to 8.
For `AggregatedStrategy`, {cluster-1:0, cluster-2: 8} is expected(scaling down the cluster with fewer replicas).
But the result now will be {cluster-1: 2, cluster-2: 6},

Reply:
Thank you so much😄. It's a nice finding👍. I didn't consider it before. A new test is added for this case, PTAL.

Reply:
cc @Poor12 @XiShanYongYe-Chang @RainbowMango for checking this.